internal/server/data/query.go

```diff
@@ -0,0 +1,120 @@
+package data
+
+import (
+	"strings"
+
+	"github.com/infrahq/infra/uid"
+)
+
+type queryBuilder struct {
+	query strings.Builder
+	Args  []interface{}
+}
+
+// Query initializes a queryBuilder and returns it populated with stmt.
+// The stmt string can generally contain any SELECT, FROM, or JOIN
+// clauses, and may contain the entire query as long as there are no
+// query parameters.
+// Use queryBuilder.B to add sections of a query with parameters.
+func Query(stmt string) *queryBuilder {
+	q := &queryBuilder{}
+	q.query.WriteString(stmt + " ")
+	return q
+}
+
+// B adds clause and args to the query. The clause must be a trusted string
+// literal. Any arguments must be passed as args so that they are properly
+// escaped by the database driver.
+func (q *queryBuilder) B(clause string, args ...interface{}) {
+	q.query.WriteString(clause + " ")
```

Comment:
Im really weary of adding this string clause directly to a query string, feels vulnerable

Reply:
We definitely need a linter for this, so that we only allow single string literals (never concatenated strings, variables, or function return values) to be passed as the first argument to `B`.

That linter will prevent any possible injection because a string literal can never come from the user, so there's no possible way for them to inject something.

All parameters from the user have to be passed as `args`.

I can work on that linter next week. It should be pretty easy to implement. We'll need to allow some exceptions for `columnsForInsert` and `placeholderForColumns` and `Table.Table`. Those exceptions are likely the most involved.

I don't see any way to avoid accepting string literals here. Did you have something in mind?

Reply:
I didn't have anything specific in mind, the linter sounds like it could work. I'll keep an eye out for that PR.

Reply:
Jeff recently [shared this post](https://blog.fuzzbuzz.io/fuzzing-go-apis-for-sql-injection/) about using fuzz testing to detect possible injection attacks. I think that's another good safeguard. I'll see if we can get some fuzz tests setup for our APIs to catch anything that might sneak past the linter.